app/src/main/java/com/codesoom/application/ProductService.java

```diff
@@ -0,0 +1,58 @@
+package com.codesoom.application;
+
+import com.codesoom.domain.Product;
+import com.codesoom.exception.ProductNotFoundException;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+@Service
+public class ProductService {
+    private final Map<Long, Product> products = new LinkedHashMap<>();
+    private Long newId = 0L;
+
+    public List<Product> getProducts() {
+        return new ArrayList<Product>(products.values());
+    }
+
+    public Optional<Product> getProduct(Long id) {
+        Optional<Product> product = findProduct(id);
+        if (!product.isPresent()) {
+            throw new ProductNotFoundException("요청한 " + id + "의 Product가 없습니다.");
+        }
+        return product;
+    }
+
+    public Product createProduct(Product product) {
+        Long id = generateId();
+        product.setId(id);
+        products.put(id, product);
+
+        return product;
+    }
+
+    public List<Product> deleteProduct(Long id) {
+        Optional<Product> product = findProduct(id);
+        if (!product.isPresent()) {
+            throw new ProductNotFoundException("요청한 " + id + "의 Product가 없습니다.");
```

Comment:
이런 예외 메시지를 작성할 때에는 디버깅하는 사람에게 도움을 준다고 생각하세요. 예외적인 상황이 발생해 어떤 작업이 좌절되었는지도 설명해주면 좋습니다. 예를 들어 다음과 같이 작성할 수 있겠죠.

```suggestion
            throw new ProductNotFoundException("요청한 " + id + "의 Product를 찾지 못했습니다. 삭제를 할 수 없습니다.");
```

Reply:
예외의 원인과 결과를 써주는 것이 좋겠군요